lib/scholar/decomposition/incremental_pca.ex

```diff
@@ -0,0 +1,294 @@
+           keep: [:whiten?],
+           containers: [
+             :components,
+             :singular_values,
+             :num_samples_seen,
+             :mean,
+             :variance,
+             :explained_variance,
+             :explained_variance_ratio
+           ]}
+  defstruct [
+    :components,
+    :singular_values,
+    :num_samples_seen,
+    :mean,
+    :variance,
+    :explained_variance,
+    :explained_variance_ratio,
+    :whiten?
+  ]
+
+  opts = [
+    num_components: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of principal components."
+    ],
+    whiten?: [
+      type: :boolean,
+      default: false,
+      doc: """
+      When true the `components` are divided by `num_samples` times `components` to ensure uncorrelated outputs with unit component-wise variances.
+
+      Whitening will remove some information from the transformed signal (the relative variance scales of the components)
+      but can sometimes improve the predictive accuracy of the downstream estimators by making data respect some hard-wired assumptions.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits an Incremental PCA model on a stream of batches.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return values
+
+  The function returns a struct with the following parameters:
+
+    * `:num_components` - The number of principal components.
+
+    * `:components` - Principal axes in feature space, representing the directions of maximum variance in the data.
+      Equivalently, the right singular vectors of the centered input data, parallel to its eigenvectors.
+      The components are sorted by decreasing `:explained_variance`.
+
+    * `:singular_values` - The singular values corresponding to each of the selected components.
+      The singular values are equal to the 2-norms of the `:num_components` variables in the lower-dimensional space.
+
+    * `:num_samples_seen` - The number of data samples processed.
+
+    * `:mean` - Per-feature empirical mean.
+
+    * `:variance` - Per-feature empirical variance.
+
+    * `:explained_variance` - Variance explained by each of the selected components.
+
+    * `:explained_variance_ratio` - Percentage of variance explained by each of the selected components.
+
+    * `:whiten?` - Whether to apply whitening.
+
+  ## Examples
+
+      iex> batches = Scidata.Iris.download() |> elem(0) |> Nx.tensor() |> Nx.to_batched(10)
+      iex> ipca = Scholar.Decomposition.IncrementalPCA.fit(batches, num_components: 2)
+      iex> ipca.components
+      iex> ipca.singular_values
+  """
+  def fit(batches = %Stream{}, opts) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    IO.puts("num_components: #{opts[:num_components]}")
+
+    Enum.reduce(
+      batches,
+      nil,
+      fn batch, model -> fit_batch(model, batch, opts) end
+    )
+  end
+
+  defp fit_batch(nil, batch, opts), do: fit_head_n(batch, opts)
```

Comment:
Using `fit_head/2` instead of `fit_head_n/2` over here won't work. I think it has to do with the compilation. For some reason the options won't be passed to `fit_head/2` and `opts[:num_components]` will be `nil`. I don't know how to fix this.

Reply:
There is a typo in the option name down below, that's probably why: num_components vs num_componenets.

Reply:
Damn, good catch. :)